qiime/beta_diversity.py

```diff
@@ -198,32 +197,28 @@ def single_file_beta(input_path, metrics, tree_path, output_dir,
                         metric_f.__name__ == 'binary_dist_chisq':
                     warnings.warn('dissimilarity ' + metric_f.__name__ +
                                   ' is not parallelized, calculating the whole matrix...')
+                    otumtx = otu_table.matrix_data.T.toarray()
                     row_dissims.append(metric_f(otumtx)[rowidx])
                 else:
-                    try:
-                        row_metric = get_phylogenetic_row_metric(metric)
-                    except AttributeError:
-                        # do element by element
-                        dissims = []
-                        sample_ids = otu_table.ids()
-                        observation_ids = otu_table.ids(axis='observation')
-                        for i in range(len(sample_ids)):
-                            if is_phylogenetic:
-                                dissim = metric_f(
-                                    otumtx[[rowidx, i], :], observation_ids,
-                                    tree, [sample_ids[rowidx], sample_ids[i]],
-                                    make_subtree=(not full_tree))[0, 1]
-                            else:
-                                dissim = metric_f(otumtx[[rowidx, i], :])[0, 1]
-                            dissims.append(dissim)
-                        row_dissims.append(dissims)
-                    else:
-                        # do whole row at once
-                        dissims = row_metric(otumtx,
-                                             otu_table.ids(axis='observation'),
-                                             tree, otu_table.ids(), rowid,
-                                             make_subtree=(not full_tree))
-                        row_dissims.append(dissims)
+                    # do element by element
```

Comment:
@wasade, I think this might be it thanks for having a look, I started it
and its been running for a while in 8GB, while before I needed to use
64GB. :beer: :+1:

On (Jul-23-15|11:42), Daniel McDonald wrote:
> > -                                dissim = metric_f(
> > -                                    otumtx[[rowidx, i], :], observation_ids,
> > -                                    tree, [sample_ids[rowidx], sample_ids[i]],
> > -                                    make_subtree=(not full_tree))[0, 1]
> > -                            else:
> > -                                dissim = metric_f(otumtx[[rowidx, i], :])[0, 1]
> > -                            dissims.append(dissim)
> > -                        row_dissims.append(dissims)
> > -                    else:
> > -                        # do whole row at once
> > -                        dissims = row_metric(otumtx,
> > -                                             otu_table.ids(axis='observation'),
> > -                                             tree, otu_table.ids(), rowid,
> > -                                             make_subtree=(not full_tree))
> > -                        row_dissims.append(dissims)
> > +                    # do element by element
> 
> @ElDeveloper, wanna give this another spin? I verified with `pdb` what branches were used and output by md5 sum is identical with and without this change on the test data.
> 
> @gregcaporaso, I'm really not sure if this change is safe for all the metrics, but for some reason, `dist_unweighted_unifrac` and `dist_weighted_unifrac` were not being treated like row metrics and instead operating on the full table because of the try/except. The comment here "do element by element" is actually misleading as best I can tell as the original code looked like it was doing it row by row. So I think this is correct?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/biocore/qiime/pull/2040/files#r35356369